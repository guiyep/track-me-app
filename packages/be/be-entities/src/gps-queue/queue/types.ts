export type SqsAttributesMessage = {
  userId: string;
  sessionId: string;
  data: string;
};
