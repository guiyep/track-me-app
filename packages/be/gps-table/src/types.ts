export type CellularGeneration = '3g' | '4g' | '5g';

export type GpsInfo = {
  latitude: number;
  longitude: number;
  altitude: number;
  accuracy: number;
  altitudeAccuracy: number;
  heading: number;
  speed: number;
};

export type WifiSignalInfo = {
  type: 'wifi';
  isConnected: boolean;
  strength: number;
  frequency: number;
};

export type CellularSignalInfo = {
  type: 'cellular';
  isConnected: boolean;
  isInternetReachable: boolean;
  cellularGeneration: CellularGeneration;
  carrier: string;
  isConnectionExpensive: boolean;
  strength: number;
};

export type BatteryState = 'charging' | 'discharging' | 'full';
export type ChargingSource = 'AC' | 'USB' | 'wireless' | 'unknown';
export type BatteryHealth = 'good' | 'poor' | 'unknown';

export type BatteryInfo = {
  batteryLevel: number;
  batteryState: BatteryState;
  chargingSource?: ChargingSource;
  chargingStatus: boolean;
  batteryHealth?: BatteryHealth;
};

export type GpsTableLocation = {
  lat: number;
  long: number;
  gpsInfo: GpsInfo;
  signalInfo: WifiSignalInfo | CellularSignalInfo;
  batteryInfo: BatteryInfo;
};

export type GpsTableIdentifiers = {
  email: string;
  sessionId: string;
};

export type GpsTableTime = {
  created: number;
  lastUpdated: number;
};

export type GpsTableInfo = GpsTableIdentifiers & GpsTableLocation;

export type GpsTableData = GpsTableIdentifiers &
  GpsTableLocation &
  GpsTableTime;

export type GpsTableSchema = {
  partitionKey: string; // [email]
  sortKey: `sessionId/${string}/created/${number}`; // sessionId/[sessionId]/created/[created]
  data: GpsTableData;
};

export type GpsTableLatestSessionSchema = {
  partitionKey: string; // [email]
  sortKey: 'latestSession';
  data: {
    sessionId: string;
  };
};
