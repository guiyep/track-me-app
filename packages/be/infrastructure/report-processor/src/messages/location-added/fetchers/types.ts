type CurrentWeatherUnits = {
  time: string; // "iso8601"
  interval: string; // "seconds"
  temperature: string; // "°C"
  windspeed: string; // "km/h"
  winddirection: string; // "°"
  is_day: string; // ""
  weathercode: string; // "wmo code"
};

type CurrentWeather = {
  time: string; // e.g., "2025-05-14T10:30"
  interval: number; // e.g., 900
  temperature: number; // e.g., 28.2
  windspeed: number; // e.g., 14.9
  winddirection: number; // e.g., 172
  is_day: number; // e.g., 1
  weathercode: number; // e.g., 2
};

export type WeatherFetchData = {
  latitude: number; // e.g., 1
  longitude: number; // e.g., 1
  generationtime_ms: number; // e.g., 0.026941299438476562
  utc_offset_seconds: number; // e.g., 0
  timezone: string; // "GMT"
  timezone_abbreviation: string; // "GMT"
  elevation: number; // e.g., 0
  current_weather_units: CurrentWeatherUnits;
  current_weather: CurrentWeather;
};

type LocationAddress = {
  building?: string;
  house_number?: string;
  road?: string;
  neighbourhood?: string;
  suburb?: string;
  borough?: string;
  city?: string;
  municipality?: string;
  state?: string;
  postcode?: string;
  country?: string;
  country_code?: string;
};

export type LocationFetchData = {
  place_id: number;
  licence: string;
  osm_type: string;
  osm_id: number;
  lat: string;
  lon: string;
  class: string;
  type: string;
  place_rank: number;
  importance: number;
  addresstype: string;
  name?: string;
  display_name: string;
  address: LocationAddress;
  boundingbox: [string, string, string, string];
  error?: string;
};
