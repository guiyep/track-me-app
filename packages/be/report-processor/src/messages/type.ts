export type FetcherFunction<T, K> = (data: T) => Promise<K>;

export type MessageHandler<T> = (data: T) => Promise<void>;
