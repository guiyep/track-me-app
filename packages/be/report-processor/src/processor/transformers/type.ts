import type { GpsLocation } from '@track-me-app/be-entities';
import type { ReportTableLocation } from '@track-me-app/report-table';

type ReportTableLocationKey = keyof ReportTableLocation;

export type FetcherFunction<T> = (
  location: GpsLocation.Entity,
) => Promise<[ReportTableLocationKey, T]>;
