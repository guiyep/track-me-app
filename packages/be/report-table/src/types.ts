import type {
  GpsInfo,
  WifiSignalInfo,
  CellularSignalInfo,
  BatteryState,
  ChargingSource,
  BatteryHealth,
  UserIdentity,
} from '@track-me-app/gps-table';

export type BatteryInfo = {
  batteryLevel: number;
  batteryState: BatteryState;
  chargingSource?: ChargingSource;
  chargingStatus: boolean;
  batteryHealth?: BatteryHealth;
};

export type WeatherInfo = {
  temperature: number;
  humidity: number;
  pressure: number;
};

export type LocationInfo = {
  city: string;
  country: string;
  region: string;
};

export type ReportTableTime = {
  created: number;
  lastUpdated: number;
};

export type TimeInfo = {
  sentDate: number;
};

export type ReportTableLocation = {
  lat: number;
  long: number;
  gpsInfo: GpsInfo;
  signalInfo: WifiSignalInfo | CellularSignalInfo;
  batteryInfo: BatteryInfo;
  userInfo?: UserInfo;
  weatherInfo?: WeatherInfo;
  locationInfo?: LocationInfo;
  timeInfo?: TimeInfo;
};

export type UserInfo = UserIdentity;

export type ReportTableIdentifiers = {
  userId: string;
  sessionId: string;
};

export type ReportTableData = ReportTableIdentifiers &
  ReportTableLocation &
  ReportTableTime;

export type ReportTableSchema = {
  partitionKey: string; // [sessionId]
  sortKey: string; // tracking:[email1],[email2],[email3],...
  data: ReportTableData;
};
