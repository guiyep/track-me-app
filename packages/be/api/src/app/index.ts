import express from 'express';
import routes from '../routes';

const app = express();

app.use(express.json());

routes.forEach((route) => {
  app.use(route);
});

app.get('/', function (_, res) {
  res.json({ message: 'Hello World test 2' });
});

export default app;
