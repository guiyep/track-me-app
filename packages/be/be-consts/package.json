{
  "name": "@track-me-app/be-consts",
  "version": "0.1.0",
  "main": "src/index.ts",
  "devDependencies": {
    "jest": "^29.6.1",
    "@types/jest": "^29.5.3",
    "typescript": "5.1.6"
  },
  "dependencies": {}
}
