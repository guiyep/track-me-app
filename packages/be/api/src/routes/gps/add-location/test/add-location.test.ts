import { App } from 'supertest/types';
import app from '../../../../app/index';
import request from 'supertest';
import { mockClient } from 'aws-sdk-client-mock';
import { DynamoDBClient, PutItemCommand } from '@aws-sdk/client-dynamodb';
import { SQSClient, SendMessageCommand } from '@aws-sdk/client-sqs';
import type { GpsLocationPostResponseBody } from '../index';
import { getConstants } from '@track-me-app/be-consts';
import {
  generateGpsInfo,
  generateWifiSignalInfo,
  generateBatteryInfo,
} from '@track-me-app/testing';

const Consts = getConstants();

describe('POST /v1/gps/add-location/:userId/:sessionId (add location) ', () => {
  test('to return 400 on empty body', async () => {
    const response: request.Response = await request(app as App)
      .post('/v1/gps/add-location/guiyep/222gh')
      .set('Accept', 'application/json');

    expect(response.headers['content-type']).toMatch(/json/);
    expect(response.status).toEqual(400);
  });

  test('to return 400 when validation do not pass', async () => {
    const response: request.Response = await request(app as App)
      .post('/v1/gps/add-location/guiyep/222gh')
      .set('Accept', 'application/json')
      .set('Content-Type', 'application/json')
      .send({ displayName: 222 });

    expect(response.headers['content-type']).toMatch(/json/);
    expect(response.status).toEqual(400);
  });

  test('to return 200 when correctly called', async () => {
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const dynamoMockClient = mockClient(DynamoDBClient);
    dynamoMockClient
      .on(PutItemCommand, {
        TableName: Consts.GpsTable.TABLE_NAME,
      })
      .resolves({});

    const snsMockClient = mockClient(SQSClient);
    snsMockClient
      .on(SendMessageCommand, {
        QueueUrl: Consts.GpsLocationsQueue.QUEUE_URL,
        MessageBody: Consts.GpsLocationsQueue.GPS_LOCATION_ADDED_COMMAND,
      })
      .resolves({ MessageId: '222' });

    const response: request.Response = await request(app as App)
      .post('/v1/gps/add-location/guiyep/222gh')
      .set('Accept', 'application/json')
      .set('Content-Type', 'application/json')
      .send({
        lat: 2,
        long: 2,
        gpsInfo: generateGpsInfo(1),
        signalInfo: generateWifiSignalInfo(1),
        batteryInfo: generateBatteryInfo(1),
      });

    const body = response.body as GpsLocationPostResponseBody;

    expect(response.headers['content-type']).toMatch(/json/);
    expect(response.status).toEqual(200);
    expect(body.data).toMatchSnapshot();
  });
});
