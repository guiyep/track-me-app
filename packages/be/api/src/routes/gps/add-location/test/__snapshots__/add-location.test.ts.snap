// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /gps/add-location/:email/:sessionId (add location)  to return 200 when correctly called 1`] = `
{
  "data": {
    "created": 1577836800000,
    "displayName": "asd",
    "email": "guiyep@gmail.com",
    "lastUpdated": 1577836800000,
    "lat": 2,
    "long": 2,
    "sessionId": "222gh",
  },
  "partitionKey": "guiyep@gmail.com",
  "sortKey": "sessionId:222gh/created:1577836800000",
}
`;
